File: contrib/rust-sdk/perfetto-protos-gpu/examples/gpu_counters.rs
// Copyright (C) 2025 Rivos Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use perfetto_sdk::{
    data_source::*, pb_decoder::*, producer::*, protos::trace::trace_packet::TracePacket,
};

use perfetto_sdk_protos_gpu::protos::{
    common::gpu_counter_descriptor::*, config::data_source_config::*,
    config::gpu::gpu_counter_config::*, trace::gpu::gpu_counter_event::*,
    trace::trace_packet::prelude::*,
};

use std::{
    error::Error,
    sync::{Arc, Mutex},
    time::{Duration, Instant},
};

#[derive(Debug)]
struct InstanceState {
    config: GpuCounterConfig,
    need_counter_descriptors: bool,
}

#[derive(Debug, Default)]
struct GpuCounterConfig {
    counter_period_ns: Option<u64>,
    counter_ids: Vec<u32>,
    instrumented_sampling: Option<bool>,
    fix_gpu_clock: Option<bool>,
}

impl GpuCounterConfig {
    fn decode(&mut self, data: &[u8]) -> &mut Self {
        use PbDecoderField::*;
        const COUNTER_PERIOD_NS_ID: u32 = GpuCounterConfigFieldNumber::CounterPeriodNs as u32;
        const COUNTER_IDS_ID: u32 = GpuCounterConfigFieldNumber::CounterIds as u32;
        const INSTRUMENTED_SAMPLING_ID: u32 =
            GpuCounterConfigFieldNumber::InstrumentedSampling as u32;
        const FIX_GPU_CLOCK_ID: u32 = GpuCounterConfigFieldNumber::FixGpuClock as u32;
        for item in PbDecoder::new(data) {
            match item.as_ref().unwrap_or_else(|e| panic!("Error: {}", e)) {
                (COUNTER_PERIOD_NS_ID, Varint(v)) => self.counter_period_ns = Some(*v),
                (COUNTER_IDS_ID, Varint(v)) => self.counter_ids.push(*v as u32),
                (INSTRUMENTED_SAMPLING_ID, Varint(v)) => self.instrumented_sampling = Some(*v != 0),
                (FIX_GPU_CLOCK_ID, Varint(v)) => self.fix_gpu_clock = Some(*v != 0),
                _ => println!("WARNING: unknown GpuCounterConfig field: {:?}", item),
            }
        }
        self
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    const GPU_COUNTER_CONFIG_ID: u32 = DataSourceConfigExtFieldNumber::GpuCounterConfig as u32;
    let producer_args = ProducerInitArgsBuilder::new().backends(Backends::SYSTEM);
    Producer::init(producer_args.build());
    let mut data_source = DataSource::new();
    let instances: Arc<Mutex<[Option<InstanceState>; 8]>> =
        Arc::new(Mutex::new([None, None, None, None, None, None, None, None]));
    let instances_for_setup = Arc::clone(&instances);
    let instances_for_start = Arc::clone(&instances);
    let instances_for_stop = Arc::clone(&instances);
    let data_source_args = DataSourceArgsBuilder::new()
        .on_setup(move |inst_id, config, _args| {
            let mut instances = instances_for_setup.lock().unwrap();
            let mut gpu_counter_config = GpuCounterConfig::default();
            for item in PbDecoder::new(config) {
                if let (GPU_COUNTER_CONFIG_ID, PbDecoderField::Delimited(value)) =
                    item.unwrap_or_else(|e| panic!("Error: {}", e))
                {
                    gpu_counter_config.decode(value);
                }
            }
            instances[inst_id as usize] = Some(InstanceState {
                config: gpu_counter_config,
                need_counter_descriptors: false,
            });
            println!(
                "OnSetup id: {} config: {:?}",
                inst_id,
                instances[inst_id as usize].as_ref().unwrap().config
            );
        })
        .on_start(move |inst_id, _| {
            let mut instances = instances_for_start.lock().unwrap();
            if let Some(state) = instances[inst_id as usize].as_mut() {
                state.need_counter_descriptors = true;
            }
            println!("OnStart id: {}", inst_id);
        })
        .on_stop(move |inst_id, _| {
            let mut instances = instances_for_stop.lock().unwrap();
            if let Some(state) = instances[inst_id as usize].as_mut() {
                state.need_counter_descriptors = false;
            }
            println!("OnStop id: {}", inst_id);
        });
    let start_time = Instant::now();
    data_source.register("gpu.counters.example", data_source_args.build())?;
    loop {
        data_source.trace(|ctx: &mut TraceContext| {
            // Fixed set of counters: sin, cos, tan.
            const COUNTER_IDS: [u32; 3] = [1, 2, 3];
            let inst_id = ctx.instance_index();
            let elapsed_secs = start_time.elapsed().as_secs_f64();
            let need_descriptors = {
                let mut instances = instances.lock().unwrap();
                match instances[inst_id as usize].as_mut() {
                    Some(state) => std::mem::replace(&mut state.need_counter_descriptors, false),
                    None => false,
                }
            };
            ctx.add_packet(|packet: &mut TracePacket| {
                packet.set_gpu_counter_event(|event: &mut GpuCounterEvent| {
                    for i in COUNTER_IDS.iter() {
                        event.set_counters(|counter: &mut GpuCounter| {
                            counter.set_counter_id(*i);
                            match i {
                                1 => counter.set_double_value(elapsed_secs.sin()),
                                2 => counter.set_double_value(elapsed_secs.cos()),
                                _ => counter.set_double_value(elapsed_secs.tan()),
                            };
                        });
                    }
                    if need_descriptors {
                        event.set_counter_descriptor(|desc: &mut GpuCounterDescriptor| {
                            for i in COUNTER_IDS.iter() {
                                desc.set_specs(|desc: &mut GpuCounterSpec| {
                                    desc.set_counter_id(*i);
                                    match i {
                                        1 => desc.set_name("sin"),
                                        2 => desc.set_name("cos"),
                                        _ => desc.set_name("tan"),
                                    };
                                });
                            }
                        });
                    }
                });
            });
        });
        let counter_period = instances
            .lock()
            .unwrap()
            .iter()
            .filter_map(|s| s.as_ref()?.config.counter_period_ns)
            .map(Duration::from_nanos)
            .min()
            .unwrap_or(Duration::from_secs(1));
        std::thread::sleep(counter_period);
    }
}

File: test/trace_processor/diff_tests/parser/graphics/gpu_counters.py
#!/usr/bin/env python3
# Copyright (C) 2019 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from os import sys, path

import synth_common

# See gpu_counter_descriptor.proto
NONE = 0
BYTE = 7
HERTZ = 13
MILLISECOND = 21
SECOND = 22
VERTEX = 25
PIXEL = 26
TRIANGLE = 27

trace = synth_common.create_trace()

# Add counter specs.
trace.add_gpu_counter_spec(
    ts=1,
    gpu_id=0,
    counter_id=31,
    name="Vertex / Second",
    description="Number of vertices per second",
    unit_numerators=[VERTEX],
    unit_denominators=[SECOND])
trace.add_gpu_counter_spec(
    ts=2,
    gpu_id=0,
    counter_id=32,
    name="Fragment / Second",
    description="Number of fragments per second",
    unit_numerators=[PIXEL],
    unit_denominators=[SECOND])
trace.add_gpu_counter_spec(
    ts=3,
    gpu_id=1,
    counter_id=34,
    name="Triangle Acceleration",
    description="Number of triangles per ms-ms",
    unit_numerators=[TRIANGLE],
    unit_denominators=[MILLISECOND, MILLISECOND])
# Counter with NONE denominator (should be skipped).
trace.add_gpu_counter_spec(
    ts=4,
    gpu_id=0,
    counter_id=35,
    name="Bytes Only",
    description="Counter with NONE denominator",
    unit_numerators=[BYTE],
    unit_denominators=[NONE])
# Counter with numerator only, no denominator.
trace.add_gpu_counter_spec(
    ts=5,
    gpu_id=0,
    counter_id=36,
    name="Frequency",
    description="Counter with numerator only",
    unit_numerators=[HERTZ],
    unit_denominators=[])

# Add some counter value events.
trace.add_gpu_counter(11, 31, 5)
trace.add_gpu_counter(21, 31, 10)
trace.add_gpu_counter(31, 31, 15.5)

trace.add_gpu_counter(12, 32, 7)
trace.add_gpu_counter(22, 32, 14)
trace.add_gpu_counter(32, 32, 21)

# Counter without a spec.
trace.add_gpu_counter(13, 33, 8)
trace.add_gpu_counter(23, 33, 16)
trace.add_gpu_counter(33, 33, 25)

trace.add_gpu_counter(14, 34, 0)
trace.add_gpu_counter(24, 34, 9)
trace.add_gpu_counter(34, 34, 7)

trace.add_gpu_counter(15, 35, 1024)
trace.add_gpu_counter(25, 35, 0)

trace.add_gpu_counter(16, 36, 60)
trace.add_gpu_counter(26, 36, 0)

sys.stdout.buffer.write(trace.trace.SerializeToString())
